feat: add var automatic_channel_upgrade

the-technat opened the pull request:
Since AKS now supports automatic upgrades, we @swisspost like to use this feature. I addressed the concerns described in the issue with the following tests:

- Created cluster on `v1.23.8` using `examples/startup`.
- Configured `automatic_channel_upgrade=patch` using Terraform
- Manually upgrade cluster to `v1.23.12` using Azure CLI
- Rerun Terraform

=> As expected Terraform will downgrade the cluster to `v1.23.8`, same must the case when automatically patching minor versions and specifying the version.

So the correct usage would be that one specifies only the minor version (e.g `1.23`) and enables automatic patch upgrade or omits the version field completely and enables automatic minor upgrades. But since Terraform variable validation can't reference other variables I can't catch misconfiguration there. But If I'm honest, it makes absolutely sense to me, that you don't specify the patch version if you want automatic upgrades.

Note: the default value must be `null` and note `none` as the Provider docs would suggest. I got errors in validation when trying to pass `none` as default value.

## Issue number

Closes #280 

## Checklist before requesting a review
- [x] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [x] I have executed pre-commit on my machine
- [x] I have passed pr-check on my machine

the-technat commented:
> Potential Breaking Changes in [07aef57](https://github.com/Azure/terraform-azurerm-aks/commit/07aef578a2cd035ab37576c9481ad2b14f19d6ce): [update] "Variables.log_analytics_solution_id.Nullable" from 'true' to '' [update] "Variables.log_analytics_workspace.Nullable" from 'true' to '' [update] "Variables.log_analytics_workspace_resource_group_name.Nullable" from 'true' to ''

This was changed due to tflint complaining that nullable is by default set to true.

zioproto commented on variables.tf:
```diff
@@ -93,6 +93,20 @@ variable "api_server_authorized_ip_ranges" {
+  type        = string
+  default     = null
+  description = "(Optional) The upgrade channel for this Kubernetes Cluster. Possible values are  `patch`, `rapid`, `node-image` and `stable`. By default automatic-upgrades are turned off. See [the documentation](https://learn.microsoft.com/en-us/azure/aks/auto-upgrade-cluste) for more information"
+  validation {
```
You should use the function `contains` here:

```
validation {
    condition = (
      var.automatic_channel_upgrade == null || (
        contains(["patch", "stable", "rapid", "node-image"], var.automatic_channel_upgrade)
      )
    )
    error_message = "`automatic_channel_upgrade`'s possible values are `patch`, `stable`, `rapid` or `node-image`."
  }
```

zioproto replied:
Commenting on the tests have done earlier, would it work to validate like this ?
```
validation {
  condition = (
    var.automatic_channel_upgrade == null || (
      contains(["patch", "stable", "rapid", "node-image"], var.automatic_channel_upgrade) &&
      matches("^(?:[0-9]\\.){2}[0-9]$", var.kubernetes_version) == false
    )
  )
  error_message = "`automatic_channel_upgrade`'s possible values are `patch`, `stable`, `rapid` or `node-image` and `kubernetes_version` must not include a patch version number."
}
```

zioproto replied:
Should the node-image channel be independent from the Kubernetes patch version check ?

zioproto replied:
Hey sorry I totally missed that you cannot validate a variable referencing another variable value. But you could move this logic to the `azurerm_kubernetes_cluster` resource to validate at runtime.

the-technat replied:
@zioproto I've used `contains` in the variable validation (to get fast feedback if the value is valid) and added a second more extensive precondition to the `azurerm_kubernetes_cluster` resource. Could you have a look at it?

the-technat replied:
Resolving this for now since it's outdated with the changes suggested by @lonegunmanb.

zioproto commented on variables.tf:
```diff
@@ -232,7 +246,6 @@ variable "log_analytics_solution_id" {
   type        = string
   description = "(Optional) Existing azurerm_log_analytics_solution ID. Providing ID disables creation of azurerm_log_analytics_solution."
   default     = null
-  nullable    = true
```
I understand you removed this line because of a `tflint` error. But we should fix the CI in this case. Can you revert these 3 changes where you remove the lines `nullable = true` from the variables that are not in scope with this PR ? thanks

zioproto commented on README.md:
```diff
@@ -284,6 +284,7 @@ No modules.
 | <a name="input_agents_tags"></a> [agents\_tags](#input\_agents\_tags)                                                                                                         | (Optional) A mapping of tags to assign to the Node Pool.                                                                                                                                                                      ...
 | <a name="input_agents_type"></a> [agents\_type](#input\_agents\_type)                                                                                                         | (Optional) The type of Node Pool which should be created. Possible values are AvailabilitySet and VirtualMachineScaleSets. Defaults to VirtualMachineScaleSets.                                                               ...
 | <a name="input_api_server_authorized_ip_ranges"></a> [api\_server\_authorized\_ip\_ranges](#input\_api\_server\_authorized\_ip\_ranges)                                       | (Optional) The IP ranges to allow for incoming traffic to the server nodes.                                                                                                                                                   ...
+| <a name="input_automatic_channel_upgrade"></a> [automatic\_channel\_upgrade](#input\_automatic\_channel\_upgrade)                                                             | (Optional) The upgrade channel for this Kubernetes Cluster. Possible values are  `patch`, `rapid`, `node-image` and `stable`. By default automatic-upgrades are turned off. See [the documentation](https://learn.microsoft.c ...
```
The link to the documentation is wrong: you are missing a `r`
https://learn.microsoft.com/en-us/azure/aks/auto-upgrade-cluste

zioproto replied:
Could you also document that using this variable requires that you cannot specify a patch version in `kubernetes_version` ?

the-technat replied:
Sure, added a sentence about this and fixed the typo.

the-technat commented:
Hey @zioproto 

Thanks for your review. I'll take a look at the requested changes. Might get tomorrow until I've time.

lonegunmanb commented on main.tf:
```diff
@@ -220,6 +221,21 @@ resource "azurerm_kubernetes_cluster" "main" {
+        )
+      )
+      error_message = "Either disable automatic upgrades, only specify up to the minor version when using `automatic_channel_upgrade=patch` or don't specify `kubernetes_version` at all when using `automatic_channel_upgrade=stable|rapid|node-image`."
+
```
Would you please remove this blank line?

lonegunmanb commented on variables.tf:
```diff
@@ -93,6 +93,20 @@ variable "api_server_authorized_ip_ranges" {
+variable "automatic_channel_upgrade" {
+  type        = string
+  default     = null
+  description = "(Optional) The upgrade channel for this Kubernetes Cluster. Possible values are  `patch`, `rapid`, `node-image` and `stable`. By default automatic-upgrades are turned off. Note that you cannot use the `patch` upgrade channel and still specify the patch version using `kubernetes_version`. See [the documentation](https://learn.microsoft.com/en-us/azure/aks/auto-upgrade-cluster) for ...
```
I think we've got an extra space here:

```text
Possible values are  `patch`
```

Should be:

```text
Possible values are `patch`
```

lonegunmanb commented on variables.tf:
```diff
@@ -93,6 +93,20 @@ variable "api_server_authorized_ip_ranges" {
+  description = "(Optional) The upgrade channel for this Kubernetes Cluster. Possible values are  `patch`, `rapid`, `node-image` and `stable`. By default automatic-upgrades are turned off. Note that you cannot use the `patch` upgrade channel and still specify the patch version using `kubernetes_version`. See [the documentation](https://learn.microsoft.com/en-us/azure/aks/auto-upgrade-cluster) for ...
+  validation {
+    condition = (
+      var.automatic_channel_upgrade == null || (
```
Since the [Terraform's binary operator is not short-circuit](https://github.com/hashicorp/terraform/issues/24128), we would have an issue when `var.automatic_channel_upgrade` is `null`, though `var.automatic_channel_upgrade == null` is `true`, Terraform would continue to evaluate the value of `contains(["patch", "stable", "rapid", "node-image"], var.automatic_channel_upgrade)`, which equals to `contains(["patch", "stable", "rapid", "node-image"], null)`, and that would caused an error, the `contains` function dosen't accept `null` as parameter.

I would recommend the following expression:

```hcl
var.automatic_channel_upgrade == null ? true : (contains(["patch", "stable", "rapid", "node-image"], var.automatic_channel_upgrade))
```

the-technat replied:
Good point, I've read about this and then forgot it again. Switched to your snippet.

lonegunmanb commented on main.tf:
```diff
@@ -220,6 +221,21 @@ resource "azurerm_kubernetes_cluster" "main" {
     }
+    precondition {
+      // either we have automatic upgrades disabled
+      condition = (var.automatic_channel_upgrade == null || (
```
The same [short-circuit](https://github.com/hashicorp/terraform/issues/24128) issue we've got in `variable`'s `validation` block.

the-technat replied:
Should be solved now.

lonegunmanb commented on main.tf:
```diff
@@ -220,6 +221,21 @@ resource "azurerm_kubernetes_cluster" "main" {
       condition     = !(var.microsoft_defender_enabled && !var.log_analytics_workspace_enabled)
       error_message = "Enabling Microsoft Defender requires that `log_analytics_workspace_enabled` be set to true."
     }
+    precondition {
```
Since the `condition` for this `precondition` is complex, I would rather to create a new `local` ,in `locals.tf` file, and add some unit tests for this condition check. Would you please refer the current unit test inside `test/unit` folder and help us by adding three unit test cases to cover this `precondition`? Thanks!

the-technat replied:
Good point. I'll move it into `locals.tf` and add a unit test.

lonegunmanb replied:
>I though about your suggestion of putting the check into locals.tf but can locals contain a precondition check?

Could we extract the expression into a local varialbe with `bool` type? Like:

```hcl
locals{
automatic_channel_upgrade_check = (var.automatic_channel_upgrade == null || (
contains(["patch"], var.automatic_channel_upgrade) &&
        regex("^[0-9]{1,}\\.[0-9]{1,}$", var.kubernetes_version)
        ) || (
        # or we have set it to rapid/stable and don't specify the version at all
        contains(["rapid", "stable", "node-image"], var.automatic_channel_upgrade) &&
        var.kubernetes_version == null
        )
      )
}
```

Then the `precondition` could be:

```hcl
precondition {
      // either we have automatic upgrades disabled
      condition = local.automatic_channel_upgrade_check
...
```

the-technat replied:
I've moved the condition into a bool var.

lonegunmanb commented on main.tf:
```diff
@@ -220,6 +221,21 @@ resource "azurerm_kubernetes_cluster" "main" {
+      condition = (var.automatic_channel_upgrade == null || (
+        // or we have set it to patch and only specify the minor version
+        contains(["patch"], var.automatic_channel_upgrade) &&
+        matches("^[0-9]{1,}\\.[0-9]{1,}$", var.kubernetes_version)
```
Is there any special reason we should use this regex here instead of determining whether `var.kubernetes_version` is `null` or empty?

the-technat replied:
Exactly. Simply checking if it's null is only viable if someone set `automatic_channel_upgrade=stable|rapid` since then AKS updates to new minor version too and noone only specifies the major version I guess. But with automatic patch upgrades we want to make sure that a valid minor version (e.g `1.24` or `1.25`) is specified.

lonegunmanb replied:
Since I'm not an aks expert, I'll leave the shoot to @zioproto , you make the call.

lonegunmanb reviewed (changes requested):
Hi @the-technat thanks for opening this pr! The `pr-check` job failed so would you please run the `pre-commit` and `pr-check` job on your machine as the [document](https://github.com/Azure/terraform-azurerm-aks#pre-commit--pr-check--test) instructed then try again?

I also recommend adding this new variable in `examples/startup`'s code so our e2e test could cover it.  Again, thanks for your contribution!

zioproto commented:
@the-technat I have to go back on the `nullable = true` removals in the `variables.tf` file. Those changes were correct. The CI was updated, and now `tflint` shows as redundant to have `nullable = true` when the default value is already null.

Please dont fix it in this PR. Could you please rebase your branch on the current `master` branch ? This will solve that issue.

the-technat commented:
@zioproto, @lonegunmanb I addressed some of your suggestions already, rebased the PR and fixed the pre-commit/pr-test check. 

I'm happy to add some unit tests as well but first we need to find the correct way of validation. @lonegunmanb suggested moving the checks into `locals.tf` cause they are quite complex. I like this idea but locals can only control the value of something, not if it get's applied or not, so what about doing the following:

- remove `automatic_channel_upgrade` variable completely 
- set the `automatic_channel_upgrade` filed on the cluster resource based on the version the user specifies
  - if `kubernetes_version=1.23.12`, set `automatic_channel_upgrade=null`
  - if `kubernetes_version=1.23`, set `automatic_channel_upgrade=patch`
  - if `kubernetes_version=null`, set `automatic_channel_upgrade=stable` (or rapid, maybe controlled via extra input var)

lonegunmanb commented:
> @zioproto, @lonegunmanb I addressed some of your suggestions already, rebased the PR and fixed the pre-commit/pr-test check.
> 
> I'm happy to add some unit tests as well but first we need to find the correct way of validation. @lonegunmanb suggested moving the checks into `locals.tf` cause they are quite complex. I like this idea but locals can only control the value of something, not if it get's applied or not, so what about doing the following:
> 
> * remove `automatic_channel_upgrade` variable completely
> * set the `automatic_channel_upgrade` filed on the cluster resource based on the version the user specifies
>   
>   * if `kubernetes_version=1.23.12`, set `automatic_channel_upgrade=null`
>   * if `kubernetes_version=1.23`, set `automatic_channel_upgrade=patch`
>   * if `kubernetes_version=null`, set `automatic_channel_upgrade=stable` (or rapid, maybe controlled via extra input var)

I think we could just extract the expression we're using now for the `condition` in the `precondition` block into a new local variable, and this expression would only referred variable, so we can mock different values to test different cases. Could we do that?

the-technat commented:
@lonegunmanb ah now I see what you meant. Yeah that should work. I'll try to do it.

the-technat commented:
@lonegunmanb I added a table-driven unit test to ensure that the upgrade check produces the right result. Is this what you expected?

the-technat commented:
Tests still fail but locally they pass.

lonegunmanb commented:
> Tests still fail but locally they pass.

No worries, I'll check the test.

lonegunmanb commented:
> Tests still fail but locally they pass.

Apology for the inconvenience @the-technat , after a quick glance your unit tests are good, the failure was caused by Checkov check. I'll keep digging.

lonegunmanb commented:
Hi @the-technat , thanks for your patience, it looks like Checkov (the policy checking tool we're using now in this pipeline) was broken, to fix the Checkov we need upgrade Checkov's version in our CI runner image, but that brought another issue, in newer Checkov they've introduced new policies for Aks resource, which this module failed to comply for now. I don't want to make this pr too complex, so I want to skip these new policies, would you please help me by adding this [config file](https://gist.github.com/lonegunmanb/1f20b4d49129f20f2ba580d8da87725e) with you pr so the Checkov would just skip these policies? The new checkov file name should be `.checkov_config.yaml`, that would fix our issue.

the-technat commented:
@lonegunmanb no problem, we all well know flaky CI tests 😉. I've added the config file as you recommended and now the tests seem to pass here as well.

lonegunmanb commented on locals.tf:
```diff
@@ -1,4 +1,15 @@
+  automatic_channel_upgrade_check = var.automatic_channel_upgrade == null ? true : (
+    contains(["patch"], var.automatic_channel_upgrade) &&
+    can(regex("^[0-9]{1,}\\.[0-9]{1,}$", var.kubernetes_version))
+    ) == true ? true : (
```
Could we omit `== true` here?

```text
) ? true : (
```

the-technat replied:
Sure, I've changed it.

lonegunmanb reviewed (changes requested):
Thanks @the-technat for your update, almost LGTM, once the e2e test passed, only a tiny change request to solve, one extra commit contains only one line change would make this pr been merged.

lonegunmanb commented:
> Based on ![image](https://user-images.githubusercontent.com/28305481/207645816-e97429fa-745e-4b08-9e33-e8df98e8a3cd.png)
> 
> I switched the version to `1.24` in e2e tests. @lonegunmanb would you rerun them?

With pleasure!

lonegunmanb reviewed (approved):
Thanks @the-technat , LGTM! 🚀
